File: get_freqs_txt.py
"""Python script to sequentially calculate frequencies of words in text files in a directory and its subdirectories
Earl Kjar Brown, ekbrown byu edu (add characters to create email)"""

import os, re, statistics
from time import time

def push_freqs_dict(freqs, wds):
    for wd in wds:
        freqs[wd] = freqs.get(wd, 0) + 1
    return freqs

def get_freqs(in_dir):
    freqs = {}
    for root, dirs, files in os.walk(in_dir):
        for file in files:
            if file.endswith("txt"):
                in_path = os.path.join(root, file)
                with open(in_path, mode="r", encoding="utf8") as infile:
                    txt = infile.read().upper()
                    wds = re.findall(r"[-'’A-Z]+", txt)
                    push_freqs_dict(freqs, wds)
    return freqs


if __name__ == "__main__":
    times = []
    for i in range(10):
        print("Working on iteration:", i+1)
        t0 = time()
        freqs = get_freqs("/Users/ekb5/Corpora/spotify-podcasts-2020/podcasts-transcripts/txt")
        t1 = time()
        times.append(t1 - t0)

    print(f"N word types: {len(freqs)}")
    print(f"N word tokens: {sum(v for v in freqs.values())}")
    print(f'THE = {freqs["THE"]}, TREE = {freqs["TREE"]}, JUSTICE = {freqs["JUSTICE"]}, VERY = {freqs["VERY"]}, BACKPACK = {freqs["BACKPACK"]}')
    print("Times:", times)
    print("Mean time:", statistics.mean(times))
    print("Median time:", statistics.median(times))

File: get_freqs_txt.jl
using Dates
# using DataStructures

function push_freqs_dict(freqs, wds)
    for wd in wds

        # # Option 1: if else
        # if haskey(freqs, wd)
        #     freqs[wd] += 1
        # else
        #     freqs[wd] = 1
        # end

        # # Option 2:
        # try
        #     freqs[wd] += 1
        # catch KeyError
        #     freqs[wd] = 1
        # end 

        # Option 3: get function
        freqs[wd] = get(freqs, wd, 0) + 1

        # # Option 4: DefaultDict
        # freqs[wd] += 1

    end
    return freqs
end

function get_freqs(in_dir)
    cd(in_dir)
    freqs = Dict{String,Int}()
    # freqs = DefaultDict{String,Int}(0)
    count = 0
    for (root, dirs, files) in walkdir(in_dir)
        for file in files
            if endswith(file, "txt")
                count += 1
                if count > 1000
                    break
                end
                cur_file = joinpath(root, file)
                open(cur_file) do infile
                    txt = uppercase(read(infile, String))
                    wds = [m.match for m in eachmatch(r"[-'’A-Z]+", txt)]
                    push_freqs_dict(freqs, wds)
                end       
            end
        end
    end
    return freqs
end

times = []
for i in 1:2
    print("Working on iteration $i... ")
    t1 = now()
    freqs = get_freqs("/pathway/to/input/dir")
    t2 = now()
    push!(times, (t2 - t1) / Millisecond(1000))
    println("done.")
    wd = "THE"
    println("$wd: ", freqs[wd])
    println("N word types: ", length(freqs))
end
println("Times: ", times)
using Statistics
println("Mean: ", Statistics.mean(times))
println("Median: ", Statistics.median(times))